### PR TITLE
dev/core#4418 Add in Upgrade script to fix double json encoding

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
This adds in an upgrade script to fix potential double json encoding of the accepted credit cards

Before
----------------------------------------
Bad data possibly in the database

After
----------------------------------------
All good data

ping @larssandergreen @eileenmcnaughton @colemanw

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FiveSixtyFour.php [hunk 1]
```diff
@@ -33,6 +35,7 @@ public function upgrade_5_64_alpha1($rev): void {
     $this->addTask('Drop unused civicrm_action_mapping table', 'dropTable', 'civicrm_action_mapping');
     $this->addTask('Update post_URL/cancel_URL in logging tables', 'updateLogging');
     $this->addTask('Add in Everybody ACL Role option value', 'addEveryBodyAclOptionValue');
+    $this->addTask(ts('Fix double json encoding of accepted_credit_cards field in payment processor table'), 'fixDoubleEscapingPaymentProcessorCreditCards');
```
Comment (demeritcowboy) [inline CRM/Upgrade/Incremental/php/FiveSixtyFour.php:38]: `ts` is unusual here, since it's just a one-off message for admins.

Comment (seamuslee001) [reply CRM/Upgrade/Incremental/php/FiveSixtyFour.php:38]: There is some ts usage here https://github.com/civicrm/civicrm-core/blob/master/CRM/Upgrade/Incremental/php/FiveSixtyThree.php#L40 so *shrugs*

Comment (demeritcowboy) [reply CRM/Upgrade/Incremental/php/FiveSixtyFour.php:38]: Those are common strings repeated across upgrades, so are better candidates for ts.

Comment (demeritcowboy) [reply CRM/Upgrade/Incremental/php/FiveSixtyFour.php:38]: Although I'd argue they are still just for admins so shouldn't have ts either.

Comment (seamuslee001) [reply CRM/Upgrade/Incremental/php/FiveSixtyFour.php]: I've removed the ts now

**Comment by demeritcowboy** [comment]
I'm getting `TypeError: array_keys(): Argument #1 ($array) must be of type array, null given in array_keys() (line 70 of ...\CRM\Upgrade\Incremental\php\FiveSixtyFour.php).

If a pp doesn't have any selected the value is null.

**Comment by demeritcowboy** [comment]
Also calling json_decode on `$paymentProcessor['accepted_credit_cards']` errors even when it's good because the api call returns an array (or something like that - error is `TypeError: json_decode(): Argument #1 ($json) must be of type string, array given in json_decode() (line 71 of C...\CRM\Upgrade\Incremental\php\FiveSixtyFour.php)`

**Comment by seamuslee001** [comment]
thanks @demeritcowboy I found some issues with my unit test which weren't pointing out those issues, I believe I have now fixed both issues you have reported and made sure the unit test covers all scenarios

**Comment by demeritcowboy** [comment]
Hehe made you work for this one. Take the rest of the day off ... as per dave's comments...

Works for me.

**Comment by larssandergreen** [comment]
Unfortunately, it's messier than this and this doesn't actually work to fix the values that are stored in the db for me.
I think because the test uses addValue it won't capture what happens when you pass an array that contains JSON as we were doing with save records => `['accepted_credit_cards' => json_encode$creditCards]`. Then you end up with this `["{\"Visa\":\"Visa\",\"MasterCard\":\"MasterCard\"}"]`, which isn't going to be fixed by json_decode.

**Comment by seamuslee001** [comment]
@larssandergreen they should be identical addValue v saveRecords should produce identical situations

**Comment by demeritcowboy** [comment]
I also had `["{\"Visa\":\"Visa\",\"MasterCard\":\"MasterCard\"}"]` in the db, and the api.get returns it as an array, so then the `[0]` then gives the json string, which decoded for me to the right value.

But maybe we shouldn't be using api in the upgrade script, as has come up elsewhere.

**Comment by seamuslee001** [comment]
@demeritcowboy I think the API is safer now than earlier with regards to upgrades

**Comment by demeritcowboy** [comment]
Is `Civi\Api4\PaymentProcessor` a real class or is it some kind of virtual entity? It had no problem finding it during the real upgrade, but I don't see it in the tree?

**Comment by demeritcowboy** [comment]
Oh it's in the ext\civi_contribute extension, which I guess isn't enabled in the upgrade test suite.

**Comment by demeritcowboy** [comment]
So that might fail in real life if somebody didn't have that component enabled.

**Comment by seamuslee001** [comment]
@demeritcowboy oh probably because extension install tasks go last right so enabling the component extensions stuffs things up hmm

**Comment by seamuslee001** [comment]
ok @demeritcowboy have shifted to use executeQuery instead of the API in the upgrade should be right now right?

**Comment by larssandergreen** [comment]
OK, I just figured out how to make this work with the API version, but now I see there is a different SQL version. I'll give that a try.

**Comment by larssandergreen** [comment]
To make this work for me, I needed:
`$paymentProcessors = PaymentProcessor::get(FALSE)->addWhere('is_test', 'IS NOT NULL')->execute();`
and
`PaymentProcessor::update(FALSE)->addValue('accepted_credit_cards', json_decode(($paymentProcessor['accepted_credit_cards'])[0], TRUE))->addWhere('id', '=', $paymentProcessor['id'])->execute();` (adding [0])

**Comment by demeritcowboy** [comment]
@larssandergreen I think you were testing the initial version. There was a second version that had those.

But @seamuslee001 did you want to keep the `is_test IS NOT NULL AND domain_id IS NOT NULL` part? I assume those are weeding out anomalies from pp's created by api or something.

**Comment by larssandergreen** [comment]
@demeritcowboy Oh, that's frustrating, must have missed that. Working for me now anyways.

**Comment by seamuslee001** [comment]
@demeritcowboy no i don't need that those were weeding  out implicit filters that the API adds in (i.e. by default if you don't put a where in and an entity has domain_id column it will filter to the current domain and same with the is_test filtering by default to be true)

**Comment by demeritcowboy** [comment]
Ah ok got it.

So yes this works for me too.

**Pull request merged**